File: Javascript/Queue/M-649. Dota2 Senate.ts
function predictPartyVictory(senate) {
  const radiants: number[] = [];
  const dires: number[] = [];
  const n = senate.length;
  for (let i = 0; i < n; i++) {
    if (senate[i] === 'R') radiants.push(i);
    else dires.push(i);
  }
  while (radiants.length && dires.length) {
    if (radiants[0] < dires[0]) radiants.push(n + radiants[0]);
    else dires.push(n + dires[0]);
    radiants.shift();
    dires.shift();
  }
  return radiants.length > 0 ? 'Radiant' : 'Dire';
}

const testData = [
  {
    senate: 'RD',
    ans: 'Radiant',
  },
  {
    senate: 'RDD',
    ans: 'Dire',
  },
];

for (const data of testData) {
  console.log(data['ans'] === predictPartyVictory(data['senate']));
}

File: Python/Queue/M-649. Dota2 Senate.py
class Solution:
    @classmethod
    def predictPartyVictory(self, senate: str) -> str:
        radiants, dires = [], []
        n = len(senate)
        for i in range(n):
            if senate[i] == "R":
                radiants.append(i)
            else:
                dires.append(i)

        while len(radiants) and len(dires):
            if radiants[0] < dires[0]:
                radiants.append(n + radiants[0])
            else:
                dires.append(n + dires[0])
            radiants.pop(0)
            dires.pop(0)

        return "Radiant" if len(radiants) else "Dire"


testData = [
    {
        "senate": "RD",
        "ans": "Radiant",
    },
    {
        "senate": "RDD",
        "ans": "Dire",
    },
]

for data in testData:
    print(data["ans"] == Solution.predictPartyVictory(data["senate"]))
